File: packages/sdk-typescript/src/directory.ts
import type {
  AgentProfile,
  AgentRecord,
  AgentRegistration,
  AgentSearchQuery,
  BeamIdString,
  BrowseFilters,
  BrowseResult,
  DirectoryConfig,
  DirectoryStats,
  DomainVerification,
  KeyRotationResult,
  Delegation,
  Report,
  VerificationTier,
} from './types.js'

function getString(raw: Record<string, unknown>, ...keys: string[]): string | undefined {
  for (const key of keys) {
    const value = raw[key]
    if (typeof value === 'string' && value.length > 0) {
      return value
    }
  }
  return undefined
}

function getNumber(raw: Record<string, unknown>, ...keys: string[]): number | undefined {
  for (const key of keys) {
    const value = raw[key]
    if (typeof value === 'number' && Number.isFinite(value)) {
      return value
    }
  }
  return undefined
}

function getBoolean(raw: Record<string, unknown>, ...keys: string[]): boolean | undefined {
  for (const key of keys) {
    const value = raw[key]
    if (typeof value === 'boolean') {
      return value
    }
  }
  return undefined
}

function getStringArray(raw: Record<string, unknown>, ...keys: string[]): string[] {
  for (const key of keys) {
    const value = raw[key]
    if (Array.isArray(value)) {
      return value.filter((entry): entry is string => typeof entry === 'string')
    }
  }
  return []
}

function normalizeAgent(raw: Record<string, unknown>): AgentRecord {
  return {
    beamId: (raw.beamId ?? raw.beam_id) as AgentRecord['beamId'],
    did: (raw.did ?? '') as string,
    displayName: (raw.displayName ?? raw.display_name ?? '') as string,
    capabilities: (raw.capabilities ?? []) as string[],
    apiKey: (raw.apiKey ?? raw.api_key ?? undefined) as string | undefined,
    publicKey: (raw.publicKey ?? raw.public_key ?? '') as string,
    org: (raw.org ?? '') as string,
    trustScore: (raw.trustScore ?? raw.trust_score ?? 0) as number,
    verified: (raw.verified ?? false) as boolean,
    createdAt: (raw.createdAt ?? raw.created_at ?? '') as string,
    lastSeen: (raw.lastSeen ?? raw.last_seen ?? '') as string,
  }
}

function normalizeProfile(raw: Record<string, unknown>): AgentProfile {
  return {
    ...normalizeAgent(raw),
    description: getString(raw, 'description'),
    logoUrl: getString(raw, 'logoUrl', 'logo_url'),
    website: getString(raw, 'website'),
    verificationTier: getString(raw, 'verificationTier', 'verification_tier') as VerificationTier | undefined,
    verificationStatus: getString(raw, 'verificationStatus', 'verification_status') as AgentProfile['verificationStatus'],
    domain: getString(raw, 'domain'),
    intentsHandled: getNumber(raw, 'intentsHandled', 'intents_handled'),
  }
}

function normalizeStats(raw: Record<string, unknown>): DirectoryStats {
  return {
    totalAgents: getNumber(raw, 'totalAgents', 'total_agents', 'agents') ?? 0,
    verifiedAgents: getNumber(raw, 'verifiedAgents', 'verified_agents') ?? 0,
    intentsProcessed: getNumber(raw, 'intentsProcessed', 'totalIntents', 'total_intents') ?? 0,
    consumerAgents: getNumber(raw, 'consumerAgents', 'consumer_agents'),
    uptime: getNumber(raw, 'uptime', 'uptimeSeconds', 'uptime_seconds'),
    waitlistSize: getNumber(raw, 'waitlistSize', 'waitlist_size'),
    version: getString(raw, 'version'),
  }
}

function normalizeVerification(raw: Record<string, unknown>, fallbackDomain?: string): DomainVerification {
  return {
    domain: getString(raw, 'domain') ?? fallbackDomain ?? '',
    verified: getBoolean(raw, 'verified') ?? false,
    status: getString(raw, 'status'),
    tier: getString(raw, 'tier', 'verificationTier', 'verification_tier') as VerificationTier | undefined,
    txtName: getString(raw, 'txtName', 'txt_name'),
    txtValue: getString(raw, 'txtValue', 'txt_value', 'expected'),
    expected: getString(raw, 'expected'),
    records: getStringArray(raw, 'records'),
    checkedAt: getString(raw, 'checkedAt', 'checked_at'),
  }
}

function normalizeRotation(raw: Record<string, unknown>, beamId: BeamIdString, publicKey: string): KeyRotationResult {
  return {
    beamId: (getString(raw, 'beamId', 'beam_id') as BeamIdString | undefined) ?? beamId,
    publicKey: getString(raw, 'publicKey', 'public_key') ?? publicKey,
    rotatedAt: getString(raw, 'rotatedAt', 'rotated_at'),
    previousKey: getString(raw, 'previousKey', 'previous_key'),
  }
}

function normalizeDelegation(raw: Record<string, unknown>): Delegation {
  return {
    id: getString(raw, 'id'),
    sourceBeamId: (getString(raw, 'sourceBeamId', 'source_beam_id') ?? '') as BeamIdString,
    targetBeamId: (getString(raw, 'targetBeamId', 'target_beam_id') ?? '') as BeamIdString,
    scope: getString(raw, 'scope') ?? '',
    expiresAt: getString(raw, 'expiresAt', 'expires_at'),
    createdAt: getString(raw, 'createdAt', 'created_at'),
    status: getString(raw, 'status'),
  }
}

function normalizeReport(raw: Record<string, unknown>): Report {
  return {
    id: getString(raw, 'id'),
    reporterBeamId: (getString(raw, 'reporterBeamId', 'reporter_beam_id') ?? '') as BeamIdString,
    targetBeamId: (getString(raw, 'targetBeamId', 'target_beam_id') ?? '') as BeamIdString,
    reason: getString(raw, 'reason') ?? '',
    createdAt: getString(raw, 'createdAt', 'created_at'),
    status: getString(raw, 'status'),
  }
}

export class BeamDirectoryError extends Error {
  constructor(
    message: string,
    public readonly statusCode: number
  ) {
    super(message)
    this.name = 'BeamDirectoryError'
  }
}

export class BeamDirectory {
  private readonly baseUrl: string
  private readonly headers: Record<string, string>

  constructor(config: DirectoryConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, '')
    this.headers = {
      'Content-Type': 'application/json',
      ...(config.apiKey && { 'x-api-key': config.apiKey })
    }
  }

  private async request<T>(path: string, init?: RequestInit): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      ...init,
      headers: {
        ...this.headers,
        ...(init?.headers ?? {}),
      },
    })

    if (!res.ok) {
      const body = await res.json().catch(() => ({ error: res.statusText })) as { error?: string }
      throw new BeamDirectoryError(body.error ?? res.statusText, res.status)
    }

    if (res.status === 204) {
      return undefined as T
    }

    return res.json() as Promise<T>
  }

  async register(registration: AgentRegistration): Promise<AgentRecord> {
    const body = await this.request<Record<string, unknown>>('/agents/register', {
      method: 'POST',
      body: JSON.stringify(registration)
    })
    return normalizeAgent(body)
  }

  async lookup(beamId: BeamIdString): Promise<AgentRecord | null> {
    const res = await fetch(`${this.baseUrl}/agents/${encodeURIComponent(beamId)}`, {
      headers: this.headers
    })
    if (res.status === 404) return null
    if (!res.ok) throw new BeamDirectoryError(`Lookup failed: ${res.statusText}`, res.status)
    const body = await res.json() as Record<string, unknown>
    return normalizeAgent(body)
  }

  async search(query: AgentSearchQuery): Promise<AgentRecord[]> {
    const params = new URLSearchParams()
    if (query.org) params.set('org', query.org)
    if (query.capabilities?.length) params.set('capabilities', query.capabilities.join(','))
    if (query.minTrustScore !== undefined) params.set('minTrustScore', String(query.minTrustScore))
    if (query.limit !== undefined) params.set('limit', String(query.limit))

    const body = await this.request<{ agents?: Record<string, unknown>[] } | Record<string, unknown>[]>(`/agents/search?${params.toString()}`)
    const raw = Array.isArray(body) ? body : (body.agents ?? [])
    return raw.map(normalizeAgent)
  }

  async browse(page = 1, filters: BrowseFilters = {}): Promise<BrowseResult> {
    const params = new URLSearchParams({ page: String(page) })
    if (filters.capability) params.set('capability', filters.capability)
    if (filters.tier) params.set('tier', filters.tier)
    if (filters.verified_only) params.set('verified_only', String(filters.verified_only))

    try {
      const body = await this.request<Record<string, unknown>>(`/agents/browse?${params.toString()}`)
      const agentsRaw = Array.isArray(body['agents'])
        ? body['agents'].filter((entry): entry is Record<string, unknown> => typeof entry === 'object' && entry !== null)
        : []
      return {
        page: getNumber(body, 'page') ?? page,
        pageSize: getNumber(body, 'pageSize', 'page_size') ?? agentsRaw.length,
        total: getNumber(body, 'total') ?? agentsRaw.length,
        agents: agentsRaw.map(normalizeProfile),
      }
    } catch (error) {
      if (!(error instanceof BeamDirectoryError) || error.statusCode !== 404) {
        throw error
      }

      const search = await this.search({
        capabilities: filters.capability ? [filters.capability] : undefined,
        limit: 20,
      })

      const filtered = search.filter((agent) => {
        if (filters.verified_only && !agent.verified) return false
        return true
      })

      return {
        page,
        pageSize: filtered.length,
        total: filtered.length,
        agents: filtered.map((agent) => ({
          ...agent,
          verificationTier: agent.verified ? 'verified' : 'basic',
        })),
      }
    }
  }

  async getStats(): Promise<DirectoryStats> {
    const body = await this.request<Record<string, unknown>>('/stats')
    return normalizeStats(body)
  }

  async updateProfile(beamId: BeamIdString, fields: { description?: string; logo_url?: string; website?: string }): Promise<AgentProfile> {
    const body = await this.request<Record<string, unknown>>(`/agents/${encodeURIComponent(beamId)}/profile`, {
      method: 'PATCH',
      body: JSON.stringify(fields),
    })
    return normalizeProfile(body)
  }

  async verifyDomain(beamId: BeamIdString, domain: string): Promise<DomainVerification> {
    const body = await this.request<Record<string, unknown>>(`/agents/${encodeURIComponent(beamId)}/verify/domain`, {
      method: 'POST',
      body: JSON.stringify({ domain }),
    })
    return normalizeVerification(body, domain)
  }

  async checkDomainVerification(beamId: BeamIdString): Promise<DomainVerification> {
    const body = await this.request<Record<string, unknown>>(`/agents/${encodeURIComponent(beamId)}/verify/domain`)
    return normalizeVerification(body)
  }

  async rotateKeys(beamId: BeamIdString, publicKey: string, rotationProof?: string): Promise<KeyRotationResult> {
    const body = await this.request<Record<string, unknown>>(`/agents/${encodeURIComponent(beamId)}/keys/rotate`, {
      method: 'POST',
      body: JSON.stringify({
        new_public_key: publicKey,
        ...(rotationProof ? { rotation_proof: rotationProof } : {}),
      }),
    })
    return normalizeRotation(body, beamId, publicKey)
  }

  async delegate(sourceBeamId: BeamIdString, targetBeamId: BeamIdString, scope: string, expiresIn?: number): Promise<Delegation> {
    const body = await this.request<Record<string, unknown>>('/delegations', {
      method: 'POST',
      body: JSON.stringify({ sourceBeamId, targetBeamId, scope, expiresIn }),
    })
    return normalizeDelegation(body)
  }

  async report(reporterBeamId: BeamIdString, targetBeamId: BeamIdString, reason: string): Promise<Report> {
    const body = await this.request<Record<string, unknown>>('/reports', {
      method: 'POST',
      body: JSON.stringify({ reporterBeamId, targetBeamId, reason }),
    })
    return normalizeReport(body)
  }

  async heartbeat(beamId: BeamIdString): Promise<void> {
    const res = await fetch(`${this.baseUrl}/agents/${encodeURIComponent(beamId)}/heartbeat`, {
      method: 'POST',
      headers: this.headers
    })
    if (!res.ok && res.status !== 404) {
      throw new BeamDirectoryError(`Heartbeat failed: ${res.statusText}`, res.status)
    }
  }
}

File: packages/sdk-python/beam_directory/directory.py
"""BeamDirectory — HTTP client for the Beam Directory REST API."""

from __future__ import annotations

from typing import Any, Optional

import httpx

from .types import (
    AgentKeyState,
    AgentRecord,
    AgentRegistration,
    AgentSearchQuery,
    BeamIdString,
    DirectoryConfig,
    KeyRevocationResult,
    KeyRotationResult,
)


class BeamDirectoryError(Exception):
    """Raised when the directory returns an error response."""

    def __init__(self, message: str, status_code: int) -> None:
        super().__init__(message)
        self.status_code = status_code


class BeamDirectory:
    """
    HTTP client for the Beam Directory REST API.

    Usage::

        dir = BeamDirectory(DirectoryConfig(base_url="https://api.beam.directory"))
        record = await dir.register(registration)
        agent  = await dir.lookup("jarvis@yourorg.beam.directory")
        agents = await dir.search(AgentSearchQuery(org="yourorg"))
    """

    def __init__(self, config: DirectoryConfig) -> None:
        self._base_url = config.base_url.rstrip("/")
        self._headers: dict[str, str] = {"Content-Type": "application/json"}
        if config.api_key:
            self._headers["Authorization"] = f"Bearer {config.api_key}"

    # ── Registration ───────────────────────────────────────────────────────────

    async def register(self, registration: AgentRegistration) -> AgentRecord:
        """Register an agent with the directory."""
        async with httpx.AsyncClient() as client:
            res = await client.post(
                f"{self._base_url}/agents/register",
                json=registration.to_dict(),
                headers=self._headers,
                timeout=30.0,
            )
        self._raise_for_status(res, "Registration failed")
        return AgentRecord.from_dict(res.json())

    # ── Lookup ─────────────────────────────────────────────────────────────────

    async def lookup(self, beam_id: BeamIdString) -> Optional[AgentRecord]:
        """Look up an agent by Beam ID. Returns None if not found."""
        async with httpx.AsyncClient() as client:
            res = await client.get(
                f"{self._base_url}/agents/{beam_id}",
                headers=self._headers,
                timeout=10.0,
            )
        if res.status_code == 404:
            return None
        self._raise_for_status(res, "Lookup failed")
        return AgentRecord.from_dict(res.json())

    # ── Search ─────────────────────────────────────────────────────────────────

    async def search(self, query: Optional[AgentSearchQuery] = None) -> list[AgentRecord]:
        """Search agents. Returns a list of matching AgentRecords."""
        params = (query or AgentSearchQuery()).to_params()
        async with httpx.AsyncClient() as client:
            res = await client.get(
                f"{self._base_url}/agents/search",
                params=params,
                headers=self._headers,
                timeout=10.0,
            )
        self._raise_for_status(res, "Search failed")
        body: Any = res.json()
        if isinstance(body, list):
            return [AgentRecord.from_dict(a) for a in body]
        # Handle { agents: [...] } envelope
        agents = body.get("agents", [])
        return [AgentRecord.from_dict(a) for a in agents]

    # ── Heartbeat ──────────────────────────────────────────────────────────────

    async def heartbeat(self, beam_id: BeamIdString) -> None:
        """Send a heartbeat to keep the agent's lastSeen timestamp fresh."""
        async with httpx.AsyncClient() as client:
            res = await client.post(
                f"{self._base_url}/agents/{beam_id}/heartbeat",
                headers=self._headers,
                timeout=5.0,
            )
        if res.status_code not in (200, 204, 404):
            self._raise_for_status(res, "Heartbeat failed")

    async def list_keys(self, beam_id: BeamIdString) -> AgentKeyState:
        async with httpx.AsyncClient() as client:
            res = await client.get(
                f"{self._base_url}/agents/{beam_id}/keys",
                headers=self._headers,
                timeout=10.0,
            )
        self._raise_for_status(res, "List keys failed")
        body: Any = res.json()
        return AgentKeyState.from_dict(body.get("keyState") if isinstance(body, dict) else None)

    async def rotate_keys(
        self,
        beam_id: BeamIdString,
        public_key: str,
        *,
        rotation_proof: Optional[str] = None,
        signature: Optional[str] = None,
        timestamp: Optional[str] = None,
    ) -> KeyRotationResult:
        payload: dict[str, Any] = {"new_public_key": public_key}
        if rotation_proof:
            payload["rotation_proof"] = rotation_proof
        if signature:
            payload["signature"] = signature
        if timestamp:
            payload["timestamp"] = timestamp

        async with httpx.AsyncClient() as client:
            res = await client.post(
                f"{self._base_url}/agents/{beam_id}/keys/rotate",
                json=payload,
                headers=self._headers,
                timeout=30.0,
            )
        self._raise_for_status(res, "Key rotation failed")
        return KeyRotationResult.from_dict(res.json(), beam_id=beam_id, public_key=public_key)

    async def revoke_key(
        self,
        beam_id: BeamIdString,
        public_key: str,
        *,
        signature: Optional[str] = None,
        timestamp: Optional[str] = None,
    ) -> KeyRevocationResult:
        payload: dict[str, Any] = {"public_key": public_key}
        if signature:
            payload["signature"] = signature
        if timestamp:
            payload["timestamp"] = timestamp

        async with httpx.AsyncClient() as client:
            res = await client.post(
                f"{self._base_url}/agents/{beam_id}/keys/revoke",
                json=payload,
                headers=self._headers,
                timeout=30.0,
            )
        self._raise_for_status(res, "Key revocation failed")
        return KeyRevocationResult.from_dict(res.json(), beam_id=beam_id)

    # ── Private helpers ────────────────────────────────────────────────────────

    @staticmethod
    def _raise_for_status(res: httpx.Response, prefix: str) -> None:
        if not res.is_success:
            try:
                body = res.json()
                msg = body.get("error", res.text)
            except Exception:
                msg = res.text
            raise BeamDirectoryError(f"{prefix}: {msg}", res.status_code)
